prov/rxm/src/rxm_cq.c

```diff
@@ -784,7 +1003,144 @@ static ssize_t rxm_rndv_send_ack(struct rxm_rx_buf *rx_buf)
 	return ret;
 }
 
+static ssize_t rxm_rndv_send_done(struct rxm_ep *rxm_ep, struct rxm_tx_rndv_buf *tx_buf)
+{
+	struct rxm_deferred_tx_entry *def_tx_entry;
+	ssize_t ret;
+
+	if (sizeof(tx_buf->pkt) <= rxm_ep->inject_limit) {
+		ret = rxm_rndv_send_done_inject(tx_buf);
+		if (!ret)
+			goto out;
+
+		if (ret != -FI_EAGAIN) {
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"send done via inject failed for MSG provider\n");
+			return ret;
+		}
+	}
+
+	tx_buf->done_buf = rxm_tx_buf_alloc(rxm_ep, RXM_BUF_POOL_TX_DONE);
+	if (!tx_buf->done_buf) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"ran out of buffers from DONE buffer pool\n");
+		return -FI_EAGAIN;
+	}
+	assert(tx_buf->done_buf->pkt.ctrl_hdr.type == rxm_ctrl_rndv_done);
+	assert(tx_buf->hdr.state == RXM_RNDV_WRITE);
 
+	tx_buf->done_buf->pkt.ctrl_hdr.conn_id = tx_buf->conn->handle.remote_key;
+	tx_buf->done_buf->pkt.ctrl_hdr.msg_id = tx_buf->pkt.ctrl_hdr.msg_id;
+
+	ret = fi_send(tx_buf->conn->msg_ep, &tx_buf->done_buf->pkt, sizeof(tx_buf->done_buf->pkt),
+		      tx_buf->done_buf->hdr.desc, 0, tx_buf);
+	if (ret) {
+		if (ret == -FI_EAGAIN) {
+			def_tx_entry = rxm_ep_alloc_deferred_tx_entry(rxm_ep,
+					tx_buf->conn, RXM_DEFERRED_TX_RNDV_DONE);
+			if (!def_tx_entry) {
+				FI_WARN(&rxm_prov, FI_LOG_CQ, "unable to "
+					"allocate TX entry for deferred DONE\n");
+				ret = -FI_EAGAIN;
+				goto err;
+			}
+
+			def_tx_entry->rndv_done.tx_buf = tx_buf;
+			rxm_ep_enqueue_deferred_tx_queue(def_tx_entry);
+			return 0;
+		} else {
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"unable to send DONE: %zd\n", ret);
+		}
+		goto err;
+	}
+out:
+	RXM_UPDATE_STATE(FI_LOG_CQ, tx_buf, RXM_RNDV_DONE_SENT);
+	return 0;
+err:
+	ofi_buf_free(tx_buf->done_buf);
+	return ret;
+}
+
+ssize_t rxm_ctrl_rndv_write_send_ack(struct rxm_rx_buf *rx_buf)
+{
+	struct rxm_deferred_tx_entry *def_tx_entry;
+	struct rxm_rndv_hdr rndv_hdr;
+	ssize_t ret;
+
+	assert(rx_buf->conn);
+
+	rxm_rndv_hdr_init(
+		rx_buf->ep, &rndv_hdr,
+		rx_buf->recv_entry->rxm_iov.iov,
+		rx_buf->recv_entry->rxm_iov.count, rx_buf->mr);
+
+	if ((sizeof(rx_buf->pkt) + sizeof(struct rxm_rndv_hdr)) <= rx_buf->ep->inject_limit) {
+		ret = rxm_rndv_write_send_ack_inject(rx_buf, &rndv_hdr);
+		if (!ret)
+			goto out;
+
+		if (ret != -FI_EAGAIN) {
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"send ack via inject failed for MSG provider\n");
+			return ret;
+		}
+	}
+
+	rx_buf->recv_entry->rndv.tx_buf =
+		rxm_tx_buf_alloc(rx_buf->ep, RXM_BUF_POOL_TX_RNDV_WRITE_ACK);
+	if (!rx_buf->recv_entry->rndv.tx_buf) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"ran out of buffers from RNDV_WRITE ACK buffer pool\n");
+		return -FI_EAGAIN;
+	}
+	assert(rx_buf->recv_entry->rndv.tx_buf->pkt.ctrl_hdr.type ==
+	       rxm_ctrl_rndv_write_ack);
+
+	rx_buf->recv_entry->rndv.tx_buf->pkt.ctrl_hdr.conn_id =
+		rx_buf->conn->handle.remote_key;
+	rx_buf->recv_entry->rndv.tx_buf->pkt.ctrl_hdr.msg_id =
+		rx_buf->pkt.ctrl_hdr.msg_id;
+
+	memcpy(rx_buf->recv_entry->rndv.tx_buf->pkt.data, &rndv_hdr,
```

Comment:
Can we initialize the data in place, rather than setting up rndv_hdr only to copy it?